pepper-apis/src/test/java/org/broadinstitute/ddp/service/WorkflowServiceTest.java

```diff
@@ -360,6 +372,136 @@ public void testBuildStateResponse_activityState_providesActivityMetadata() {
+        TransactionWrapper.useTxn(handle -> {
+            Map<String, Long> actParticipatingInTransition = setupActivitiesAndTransitionBetweenThem(handle);
+            ActivityInstanceDto newInstance = insertNewInstance(handle, actParticipatingInTransition.get("activityTransitionedToId"));
+            service.suggestNextState(
+                    handle, operatorGuid, userGuid, studyGuid, new ActivityState(
+                            actParticipatingInTransition.get("activityTransitionedFromId")
+                    )
+            );
+
+            Optional<String> latestInstanceGuid = handle.attach(JdbiActivityInstance.class)
+                    .findLatestInstanceGuidByUserGuidAndActivityId(userGuid, actParticipatingInTransition.get("activityTransitionedToId"));
+            assertTrue(latestInstanceGuid.isPresent());
+            assertEquals(newInstance.getGuid(), latestInstanceGuid.get());
+            handle.rollback();
+        });
+    }
+
+    @Test
+    public void test_givenInstanceCreationTriggersAnotherEvent_whenSuggestNextStateIsCalled_thenThatEventGetsProcessed() {
+        TransactionWrapper.useTxn(handle -> {
+            Map<String, Long> actParticipatingInTransition = setupActivitiesAndTransitionBetweenThem(handle);
+            // Setting up an event that will make EventService create a new activity instance
+            FormActivityDef triggeredFormActivity = insertNewActivity(handle);
+            // activityTransitionedFrom --> activityTransitionedTo transition with a precondition expr that is always true
+            long triggerId = handle.attach(JdbiEventTrigger.class).insert(EventTriggerType.ACTIVITY_STATUS);
+            handle.attach(JdbiActivityStatusTrigger.class).insert(
+                    triggerId, actParticipatingInTransition.get("activityTransitionedToId"), InstanceStatusType.CREATED
+            );
+            EventActionDao eventActionDao = handle.attach(EventActionDao.class);
+            long actionId = eventActionDao.insertInstanceCreationAction(triggeredFormActivity.getActivityId());
+            JdbiEventConfiguration jdbiEventConfig = handle.attach(JdbiEventConfiguration.class);
+            long eventConfigurationId = jdbiEventConfig.insert(
+                    triggerId, actionId, data.getStudyId(), Instant.now().toEpochMilli(), 1, 0, null, null, false, 1
+            );
+
+            service.suggestNextState(
+                    handle, operatorGuid, userGuid, studyGuid, new ActivityState(
+                            actParticipatingInTransition.get("activityTransitionedFromId")
+                    )
+            );
+
+            Optional<String> latestInstanceGuid = handle.attach(JdbiActivityInstance.class)
+                    .findLatestInstanceGuidByUserGuidAndActivityId(userGuid, triggeredFormActivity.getActivityId());
+            assertTrue(latestInstanceGuid.isPresent());
+            handle.rollback();
+        });
+    }
+
```

Comment:
Making sure that when a missing instance is created, all dependent events get processed.